include/buffer_control_block.h

```diff
@@ -37,6 +37,12 @@ class BufferControlBlock {
    */
   bool check_size_limit(uint32_t size, uint32_t flush_size, uint32_t max_size);
 
+  // lock for controlling read/write access to the buffer
+  std::mutex RW_lock;
+
+  // lock for controlling flushing this block and its sub-tree
+  std::mutex flush_lock;
```

Comment:
so do we have a bcb for each buffer or just the children of the root?

Reply:
For each buffer. Although for non-root buffers we don't use the mutexes (potentially a future optimization could remove the non-root ones)